pkg/kubeclient/kubelet_client.go

```diff
@@ -150,59 +161,89 @@ func (client *KubeletClient) GetSummary(host string) (*stats.Summary, error) {
 	} else {
 		entry := &CacheEntry{
 			statsSummary: summary,
-			machineInfo:  nil,
 			used:         false,
 		}
 		client.cache[host] = entry
 	}
 	return summary, err
 }
 
-func (client *KubeletClient) GetMachineInfo(host string) (*cadvisorapi.MachineInfo, error) {
-	// Get the data
-	var minfo cadvisorapi.MachineInfo
-	err := client.ExecuteRequestAndGetValue(host, specPath, &minfo)
-	// Lock and check cache
+// get node single-core Frequency, in MHz
+func (client *KubeletClient) GetNodeCpuFrequency(node *v1.Node) (float64, error) {
+	ip, err := util.GetNodeIPForMonitor(node, types.KubeletSource)
+	if err != nil {
+		glog.Errorf("Failed to IP for node %s: %s", node.Name, err)
+		return 0, err
+	}
+
+	// Get value from cache if exists
 	client.cacheLock.Lock()
-	defer client.cacheLock.Unlock()
-	entry, entryPresent := client.cache[host]
+	entry, entryPresent := client.cache[ip]
+	if entryPresent {
+		if entry.nodeCpuFreq != nil {
+			nodeFreq := *entry.nodeCpuFreq
+			client.cacheLock.Unlock()
+			return nodeFreq, nil
+		}
+	}
+	client.cacheLock.Unlock()
+
+	// We could not get a valid cpu frequency from cache, discover it.
+	var nodeFreq float64
+	minfo, err := client.GetMachineInfo(ip)
+	freqDiscovered := true
 	if err != nil {
-		if entryPresent {
-			entry.used = true
-			if entry.machineInfo == nil {
-				glog.V(2).Infof("unable to retrieve machine[%s] machine info: %v. The cached value unavailable", host, err)
-				return nil, err
+		glog.Errorf("Failed to get machine[%s] cpu.frequency from kubelet: %v. Will try pod based getter.", node.Name, err)
+		nodeIsActive := util.NodeIsReady(node) && util.NodeIsSchedulable(node)
+		if nodeIsActive {
+			nodeFreq, err = client.fallbkCpuFreqGetter.GetFrequency(node.Name)
+			if err != nil {
+				glog.Errorf("Failed to get cpufreq from getter %s: %s. Default value will be used.", node.Name, err)
+				freqDiscovered = false
 			}
-			glog.V(2).Infof("unable to retrieve machine[%s] machine info: %v. Using cached value", host, err)
-			return entry.machineInfo, nil
 		} else {
-			glog.Errorf("failed to get machine[%s] machine info: %v. No cache available", host, err)
-			return &minfo, err
+			freqDiscovered = false
 		}
+	} else {
+		nodeFreq = util.MetricKiloToMega(float64(minfo.CpuFrequency))
 	}
-	// Fill in the cache
+
+	if !freqDiscovered {
+		// We probably could not discover the cpufreq by any means and default
+		// value is what we need to stick to. However, we might be able to discover
+		// a valid one in the next discovery cycle.
+
+		return client.defaultCpuFreq, nil
+	}
+
+	// We optimistically set a cpu frequency of last discovered node as default.
+	// Clusters more often then not have similar nodes.
+	client.defaultCpuFreq = nodeFreq
```

Comment:
I don't think there is any meaning to make this thread safe. Each node could be discovered in its own goroutine (say 4 cores assigned to kubeturbo discovering 4 nodes). Possible example scenario is that one node discovers its frequency and 3 others don't, where they need to fall back on default. The node that discovers it updates the default, before 2 (of the other 3) fail, so those 2 get the updated default and the third one doesn't (which probably failed earlier and picked the original default). This will stay so for this discovery, but eventually correct itself in the next.
Let me know if you think I am missing something.

Reply:
Thanks, I am ok with this.